ocs-api/main.go

```diff
@@ -349,55 +355,179 @@ func postVoucherHandler(w http.ResponseWriter, r *http.Request) {
+	for _, u := range userDirs {
+		if u.IsDir() {
+			uDir := pubKeyDirName + "/" + u.Name()
+			keyFiles, err := ioutil.ReadDir(filepath.Clean(uDir))
+			if err != nil {
+				http.Error(w, "Error reading "+uDir+" directory: "+err.Error(), http.StatusBadRequest)
+				return
+			}
+			keyFileNames := make([]string, 0) // the array of public key file names for this user that we will add
+			for _, f := range keyFiles {
+				if !f.IsDir() {
+					keyFileNames = append(keyFileNames, f.Name())
+				}
+			}
+
+			if len(keyFileNames) > 0 {
+				pubKeyMap[u.Name()] = keyFileNames // add it to the map we will return
+			}
+		}
+	}
+
+	// Send file names to client
+	httpCode := http.StatusOK
+	if len(pubKeyMap) == 0 {
+		httpCode = http.StatusNotFound
+	}
+	outils.WriteJsonResponse(httpCode, w, pubKeyMap)
+}
+
+//============= GET /api/orgs/{org-id}/keys/{key-name} =============
+// Reads/returns an already imported voucher
+func getKeyHandler(orgId, keyName string, w http.ResponseWriter, r *http.Request) {
+	outils.Verbose("GET /api/orgs/%s/keys/%s ...", orgId, keyName)
+
+	// Determine the org id to use for the device, based on various inputs
+	deviceOrgId, httpErr := getDeviceOrgId(orgId, r)
+	if httpErr != nil {
+		http.Error(w, httpErr.Error(), httpErr.Code)
+		return
+	}
+
+	authenticated, user, httpErr := outils.ExchangeAuthenticate(r, CurrentExchangeInternalUrl, deviceOrgId, OcsDbDir+"/v1/values/agent-install.crt")
+	if httpErr != nil {
 		http.Error(w, httpErr.Error(), httpErr.Code)
 		return
+	} else if !authenticated {
+		http.Error(w, "invalid exchange credentials provided", http.StatusUnauthorized)
+		return
+	}
+
+	// Verify key file is in the db
+	pubKeyDirName := OcsDbDir + "/v1/creds/publicKeys/" + deviceOrgId
+	if err := os.MkdirAll(pubKeyDirName, 0750); err != nil { // in case the sub-dir doesn't even exist yet
+		http.Error(w, "could not create directory "+pubKeyDirName+": "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+	pubKeyFileName := pubKeyDirName + "/" + user + "/" + strings.ToLower(deviceOrgId+"_"+keyName) + "_public-key.pem"
+	if !outils.PathExists(pubKeyFileName) {
+		//http.Error(w, "Public key "+keyName+" for user "+user+" not found", http.StatusNotFound)
+		w.WriteHeader(http.StatusNotFound)
+		return
 	}
 
-	w.WriteHeader(http.StatusOK)
+	// Send public key file to client
+	//w.WriteHeader(http.StatusCreated) // http.ServeFile() does this too
+	w.Header().Set("Content-Type", "application/octet-stream")
+	w.Header().Set("Content-Disposition", "attachment; filename="+keyName)
+	http.ServeFile(w, r, pubKeyFileName)
 }
-*/
 
-//============= POST /api/keys =============
-// Receives in the body json containing the arguments required to run import-owner-key-pairs.sh script. Then it creates and import key pair into master keystore. Imports them into our keystore.
+//============= DELETE /api/orgs/{org-id}/keys/{key-name} =============
+// Reads/returns an already imported voucher
```

Comment:
Just the comment 
"Deletes an already imported key pair (public and private keys)" ?

Reply:
Thx. Fixed.